File: 문제풀이/온라인 저지/백준/# 04. Bronze Ⅱ/# 02798. 블랙잭/python/main.py
def main():
    n, m = map(int, input().split())
    numbers = list(map(int, input().split()))

    answer = search(numbers, n, m)
    print(answer, end='')


def search(numbers, n, m):
    result = 0
    for i in range(0, n - 2):
        if numbers[i] >= m:
            continue
        for j in range(i + 1, n - 1):
            if numbers[i] + numbers[j] >= m:
                continue
            for k in range(j + 1, n):
                sum = numbers[i] + numbers[j] + numbers[k]
                if sum == m:
                    return sum
                if result < sum < m:
                    result = sum
    return result


if __name__ == '__main__':
    main()

File: 문제풀이/온라인 저지/백준/# 04. Bronze Ⅱ/# 02798. 블랙잭/kotlin/Main.kt
fun main() {
    val n = readInt()
    val m = readInt()

    val numbers = IntArray(n)
    for (i in 0 until n) {
        numbers[i] = readInt()
    }

    print(solution(numbers, n, m))
}

private fun solution(numbers: IntArray, n: Int, m: Int): Int {
    var max = 0
    var sum: Int
    for (i in 0 until n - 2) {
        if (numbers[i] > m) continue // 제한보다 크면 다음으로 skip
        for (j in i + 1 until n - 1) {
            if (numbers[i] + numbers[j] > m) continue // 제한보다 크면 다음으로 skip
            for (k in j + 1 until n) {
                sum = numbers[i] + numbers[j] + numbers[k]
                if (sum == m) return sum // 제한과 같으면 바로 반환
                if (sum in (max + 1) until m) max = sum // 제한보다 작으면서 기존 최댓값보다 크면 최댓값 갱신
            }
        }
    }
    return max // 최댓값 반환
}

private fun readInt(): Int {
    var value = 0
    var input: Int
    while (true) {
        input = System.`in`.read() // 한글자 읽기
        when (input) {
            10, 32 -> return value // 공백이나 개행자면 입력을 멈추고 반환
            else -> value = value * 10 + (input - 48) // 기존값에 10을 곱하고 입력값을 숫자로 변환해서 합산
        }
    }
}
